packages/server/src/api/muster/muster.controller.ts

```diff
@@ -233,22 +237,110 @@ class MusterController {
-  ) {
-    const musterComplianceRate = await getMusterComplianceByDate(req.params.orgId, req.params.unitName, req.query.isoDate);
-    res.json({ musterComplianceRate, isoDate: req.query.isoDate });
+  async getRosterMusterComplianceByDateRange(req: ApiRequest<OrgRoleParams, null, GetMusterRosterQuery>, res: Response<Paginated<Partial<MusterCompliance>>>) {
+    assertRequestQuery(req, ['fromDate', 'toDate', 'limit', 'page']);
+
+    /* Dates come in UTC timezone.
+      For example the date 2021-07-04 11:59 PM selected in a browser
+                  comes as 2021-07-05T04:59:00.000Z */
+    const fromDate: moment.Moment = moment(req.query.fromDate, 'YYYY-MM-DD');
+    const toDate: moment.Moment = moment(req.query.toDate, 'YYYY-MM-DD');
+
+    if (!fromDate.isValid() || !toDate.isValid() || fromDate > toDate) {
+      throw new BadRequestError('Invalid ISO date range.');
+    }
+
+    const rowLimit = parseInt(req.query.limit);
+    const pageNumber = parseInt(req.query.page);
+    // When unit id is missing, then data for all units are requested
+    const unitId = req.query.unitId!;
+    const orgId = req.appOrg!.id;
+
+    Log.info(`getUserMusterCompliance() for ${JSON.stringify({ orgId, unitId, fromDate, toDate })}`);
+
+    // get units for the org
+    const orgUnits = await Unit.find({
+      where: { org: orgId },
+    });
+
+    // get unit information for the org
+    const unitIds: number[] = [];
+    const unitNames: string[] = [];
+    const configsByUnitId: Map<number, object[]> = new Map<number, object[]>();
+    orgUnits.forEach(unit => {
+      unitIds.push(unit.id);
+      unitNames.push(unit.name);
+      configsByUnitId.set(unit.id, unit.musterConfiguration);
+    });
+
+    // get the roster entries based on the unit IDs for the org
+    const rosterEntries = await Roster.find({
+      relations: ['unit'],
+      where: { unit: In(unitIds) },
+    });
+
+    // get the edipis for all roster entries
+    const rosterEdipis: string[] = [];
+    rosterEntries.forEach(rosterEntry => {
+      rosterEdipis.push(rosterEntry.edipi);
+    });
+
+    // get observations for all users on roster for this org
+    // we use query builder here to get the data back in a flat manner
+    // rather than nested
+    const observations = await Observation.createQueryBuilder('observation')
+      .select('observation.edipi, observation.timestamp, observation.report_schema_id')
+      .andWhere(`observation.timestamp between :fromDate and :toDate`, { fromDate, toDate })
+      .andWhere(`observation.edipi in (:...edipis)`, { edipis: rosterEdipis })
+      .andWhere(`observation.unit in (:...unitNames)`, { unitNames })
+      .andWhere(`observation.report_schema_org = :orgId`, { orgId })
+      .getRawMany();
+
+    const observationsByEdipi = _.groupBy(observations, obs => obs.edipi);
+    const complianceRecords: MusterCompliance[] = [];
+
+    // generate "compliance records" for each user
+    rosterEntries.forEach(rosterEntry => {
```

Comment:
I think the intent of the following chunk will be clearer if you use `map` here instead of `forEach`, and it also avoids mutations...

```typescript
const complianceRecords: MusterCompliance[] = _.map(rosterEntries, rosterEntry => {...});
```